File: 1254. Number of Closed Islands/Solution.java
class Solution {
    int n, m;
    int[][] g;
    int[][] moves = { { 1, 0 }, { -1, 0 }, { 0, 1 }, { 0, -1 } };
    public int closedIsland(int[][] grid) {
        n = grid.length;
        m = grid[0].length;
        int count = 0;
        g = grid;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == 1) continue;
                if (!dfs(i, j)) count++;
            }
        }
        return count;
    }
    public boolean dfs(int x, int y) {
        boolean res = x == 0 || y == 0 || x == n - 1 || y == m - 1;
        g[x][y] = 1;
        for (int[] mo : moves) {
            int nx = x + mo[0], ny = y + mo[1];
            if (nx < 0 || ny < 0 || nx >= n || ny >= m) continue;
            if (g[nx][ny] == 1) continue;
            res |= dfs(nx, ny);
        }
        return res;
    }
}

File: 1254. Number of Closed Islands/Solution.py
class Solution:
    def closedIsland(self, grid: List[List[int]]) -> int:
        self.n, self.m = len(grid), len(grid[0])
        self.g = grid
        self.moves = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        count = 0
        for i in range(self.n):
            for j in range(self.m):
                if self.g[i][j] == 1: continue
                if not self.dfs(i, j): count += 1
        return count
    
    def dfs(self, x, y):
        res = x == 0 or y == 0 or x == self.n - 1 or y == self.m - 1
        self.g[x][y] = 1
        for i, j in self.moves:
            nx, ny = x + i, y + j
            if nx < 0 or ny < 0 or nx >= self.n or ny >= self.m: continue
            if self.g[nx][ny] == 1: continue
            res |= self.dfs(nx, ny)
        return res